Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +133,757 @@ final class ConnectionPool {
+    func close() -> EventLoopFuture<Void> {
+        return self.channel.close()
+    }
+
+    /// Sets idle timeout handler and channel inactivity listener.
+    func setIdleTimeout(timeout: TimeAmount?) {
+        _ = self.channel.pipeline.addHandler(IdleStateHandler(writeTimeout: timeout)).flatMap { _ in
+            self.channel.pipeline.addHandler(IdlePoolConnectionHandler(connection: self))
         }
+    }
 
-        var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+    /// Removes idle timeout handler and channel inactivity listener
+    func cancelIdleTimeout() -> EventLoopFuture<Void> {
+        return self.removeHandler(IdleStateHandler.self).flatMap { _ in
+            self.removeHandler(IdlePoolConnectionHandler.self)
         }
+    }
+}
 
-        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
-            switch action {
-            case .leaseConnection(let connection):
-                return connection.removeIdleConnectionHandlersForLease().flatMapError { _ in
-                    connection.closeFuture.flatMap { // We ensure close actions are run first
-                        let defaultEventLoop = self.stateLock.withLock {
-                            self.state.defaultEventLoop
-                        }
-                        return self.makeConnection(on: preference.bestEventLoop ?? defaultEventLoop)
-                    }
-                }
-            case .makeConnection(let eventLoop):
-                return self.makeConnection(on: eventLoop)
-            case .leaseFutureConnection(let futureConnection):
-                return futureConnection
-            }
+extension Connection: Hashable {
+    static func == (lhs: Connection, rhs: Connection) -> Bool {
+        return ObjectIdentifier(lhs) == ObjectIdentifier(rhs)
+    }
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(ObjectIdentifier(self))
+    }
+}
+
+/// A connection provider of `HTTP/1.1` connections with a given `Key` (host, scheme, port)
+///
+/// On top of enabling connection reuse this provider it also facilitates the creation
+/// of concurrent requests as it has built-in politeness regarding the maximum number
+/// of concurrent requests to the server.
+class HTTP1ConnectionProvider: CustomStringConvertible {
+    enum Action {
+        case lease(Connection, Waiter)
+        case create(Waiter)
+        case replace(Connection, Waiter)
+        case closeProvider
+        case park(Connection)
+        case none
+        case fail(Waiter, Error)
+        case cancel(Connection, Bool)
+        indirect case closeAnd(Connection, Action)
+        indirect case parkAnd(Connection, Action)
+    }
+
+    struct ConnectionsState {
```

Comment:
Should we maybe move the state machine to a separate file?

Reply:
good idea, done, thanks!